### ISSUE TITLE
frontend sometimes fails to detect full set of xrandr resolution options, leading to wrong resolution & jitteriness.

### ISSUE THREAD

**Issue by ijc** [MythTV/mythtv#283]
I use the setting where there are different resolutions for the UI and recording playback. Every now and then `mythfrontend` gets into a state where it only sees the current (UI) video mode (3840x2160 @ 30) instead of the full set so when watching content it fails to switch to the desired resolution/refresh rate (generally 1920x1080 @ 50 to playback 25fps content from UK Freeview HD channels). This then leads to jitter/jerkiness in the playback.

Capturing with `general:info,file:warning,playback:info` when things are going wrong I see at frontend startup:
```
Nov 13 17:13:31 iranon mythfrontend: mythfrontend[2767]: I CoreContext mythdisplay.cpp:358 (GetDesiredScreen) Display: Found screen number 0 (DP1)
Nov 13 17:13:31 iranon mythfrontend: mythfrontend[2767]: I CoreContext mythdisplay.cpp:487 (DebugScreen) Display: Using screen 'DP1' (Make: Unknown Model: Unknown)
Nov 13 17:13:31 iranon mythfrontend: mythfrontend[2767]: I CoreContext mythdisplay.cpp:489 (DebugScreen) Display: Qt screen pixel ratio: 1.00
Nov 13 17:13:31 iranon mythfrontend: mythfrontend[2767]: I CoreContext mythdisplay.cpp:491 (DebugScreen) Display: Geometry: 3840x2160+0+0 Size(Qt): 890mmx500mm
Nov 13 17:13:31 iranon mythfrontend: mythfrontend[2767]: I CoreContext mythdisplay.cpp:1038 (DebugModes) Display: Available modes:
Nov 13 17:13:31 iranon mythfrontend: mythfrontend[2767]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 3840x2160#01130.00
Nov 13 17:13:31 iranon mythfrontend: mythfrontend[2767]: N CoreContext mythdisplay.cpp:517 (Initialise) Display: Desktop video mode: 3840x2160 30.000Hz
```
The when switching to playing a recording:
```
Nov 13 17:25:53 iranon mythfrontend: mythfrontend[2767]: I CoreContext mythdisplay.cpp:671 (SwitchToVideo) Display: Trying to match best refresh rate 25.000Hz
Nov 13 17:25:53 iranon mythfrontend: mythfrontend[2767]: E CoreContext mythdisplay.cpp:685 (SwitchToVideo) Display: Trying mode 1920x1080@25.000Hz
Nov 13 17:25:53 iranon mythfrontend: mythfrontend[2767]: E CoreContext platforms/mythdisplayx11.cpp:201 (SwitchToVideoMode) DisplayX11: Desired resolution and frame rate not found.
Nov 13 17:25:53 iranon mythfrontend: mythfrontend[2767]: E CoreContext mythdisplay.cpp:690 (SwitchToVideo) Display: Failed to change mode to 1920x1080@25.000Hz
Nov 13 17:25:53 iranon mythfrontend: mythfrontend[2767]: I CoreContext mythvideoout.cpp:1029 (InitDisplayMeasurements) VideoOutput: Display aspect ratio: 1.78 (Detected)
```

When in this state running `xrandr` shows that it sees all the expected modes and I can switch to them manually etc.

When things are working well I instead see:
```
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:358 (GetDesiredScreen) Display: Found screen number 0 (DP1)
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:487 (DebugScreen) Display: Using screen 'DP1' (Make: Unknown Model: Unknown)
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:489 (DebugScreen) Display: Qt screen pixel ratio: 1.00
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:491 (DebugScreen) Display: Geometry: 3840x2160+0+0 Size(Qt): 890mmx500mm
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext platforms/mythdisplayx11.cpp:161 (GetVideoModes) DisplayX11: Ignoring interlaced mode 1920x1080 60.00i
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext platforms/mythdisplayx11.cpp:161 (GetVideoModes) DisplayX11: Ignoring interlaced mode 1920x1080 50.00i
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext platforms/mythdisplayx11.cpp:161 (GetVideoModes) DisplayX11: Ignoring interlaced mode 1920x1080 59.94i
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1038 (DebugModes) Display: Available modes:
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 4096x2160#01130.00#01129.97#01125.00#01124.00#01123.98
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 3840x2160#01130.00#01129.97#01125.00#01124.00#01123.98
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 1920x1080#01160.00#01159.94#01150.00#01130.00#01129.97#01125.00#01124.00#01123.98
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 1680x1050#01159.88
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 1600x900#01160.00
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 1440x900#01159.90
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 1366x768#01159.79
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 1280x1024#01175.02#01160.02
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 1280x800#01159.91
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 1280x720#01160.00#01159.94#01150.00
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 1152x864#01175.00
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 1024x768#01175.03#01170.07#01160.00
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 832x624#01174.55
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 800x600#01175.00#01172.19#01160.32
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 720x576#01150.00
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 720x480#01160.00#01159.94
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 720x400#01170.08
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:1047 (DebugModes) 640x480#01175.00#01172.81#01166.67#01160.00#01159.94
Nov 13 17:30:44 iranon mythfrontend: mythfrontend[2280]: N CoreContext mythdisplay.cpp:517 (Initialise) Display: Desktop video mode: 3840x2160 30.000Hz
...
Nov 13 17:31:17 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:671 (SwitchToVideo) Display: Trying to match best refresh rate 25.000Hz
Nov 13 17:31:17 iranon mythfrontend: mythfrontend[2280]: E CoreContext mythdisplay.cpp:685 (SwitchToVideo) Display: Trying mode 1920x1080@25.000Hz
Nov 13 17:31:17 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:984 (WaitForScreenChange) Display: Waiting for resolution change
Nov 13 17:31:17 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:426 (GeometryChanged) Display: New screen geometry: 1920x1080+0+0
Nov 13 17:31:17 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:402 (PhysicalDPIChanged) Display: Qt screen pixel ratio changed to 54.83
Nov 13 17:31:17 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythdisplay.cpp:701 (SwitchToVideo) Display: Switched to 1920x1080@50.000Hz for video 1920x1080
Nov 13 17:31:17 iranon mythfrontend: mythfrontend[2280]: I CoreContext mythvideoout.cpp:975 (ResizeForVideo) VideoOutput: Aspect ratio: 1.78 (Detected)
```

* **Platform**:

Debian testing/unstable hybrid (about a week since I last updated at this point), using deb-multimedia packages of MythTV

`Linux iranon 5.9.0-1-amd64 #1 SMP Debian 5.9.1-1 (2020-10-17) x86_64 GNU/Linux`

Intel UHD Graphics 630, using xorg.

* **MythTV version**:

fixes/31 at eb3c84de5f.

* **Package version**:

`31.0+fixes20201018.giteb3c84de5f-dmo1`

* **Component**:

Frontend

### What steps will reproduce the bug?

I'm unclear, it seems to happen randomly. It could be correlated with restating the frontend. Once it happens it seems no number of restarts (full X server restart) will fix the issue -- I end up having to reboot.

I initially had a suspicion that there might be a correlation with restarting the frontend while the TV (a Samsung) was turned off, but I'm not convinced that theory holds water any more based on what I'm seeing, I'm continuing to keep an eye on when that happens.

### How often does it reproduce? Is there a required condition?

See above.

### What is the expected behaviour?

Should see all the modes and refresh rates and pick the appropriate one.

### What do you see instead?

Sees only the current mode.

### Additional information

I upgraded from fixes/30 to fixes/31 on 20 September. I'm sure I wasn't seeing this with fixes/30 but I can't say for sure that it started happening immediately on update. Upgrades:

* 20-Sep 31.0+fixes20200906.gitb282809197-dmo1 (initial upgrade to 31, from 30.0+fixes20191217.gita27754ae7f-dmo2+0.hellion6 which I am pretty certain was good, this had some local patches for music stuff, nothing related to video)
* 10-Oct 31.0+fixes20200926.git8ca7747648-dmo1
* 24-Oct 31.0+fixes20201018.giteb3c84de5f-dmo1

It looks like I picked up a new `xserver-xorg` on 24-Oct, going from 1:7.7+20 to 1:7.7+21 but `xserver-xorg-video-intel` has been at version `2:2.99.917+git20200714-1` since 21-Aug which I think makes it less likely to be involved.

At time of writing Xorg related packages are up to date with Debian testing
```
mythtv@iranon:~$ dpkg -l *xorg* | grep ^ii
ii  xorg-docs-core                 1:1.7.1-1.1              all          Core documentation for the X.org X Window System
ii  xserver-xorg                   1:7.7+21                 amd64        X.Org X server
ii  xserver-xorg-core              2:1.20.8-2               amd64        Xorg X server - core server
ii  xserver-xorg-input-evdev       1:2.10.6-2               amd64        X.Org X server -- evdev input driver
ii  xserver-xorg-input-kbd         1:1.9.0-1+b2             amd64        X.Org X server -- keyboard input driver
ii  xserver-xorg-input-mouse       1:1.9.3-1                amd64        X.Org X server -- mouse input driver
ii  xserver-xorg-legacy            2:1.20.8-2               amd64        setuid root Xorg server wrapper
ii  xserver-xorg-video-intel       2:2.99.917+git20200714-1 amd64        X.Org X server -- Intel i8xx, i9xx display driver
mythtv@iranon:~$ dpkg -l *intel* | grep ^ii
ii  intel-gpu-tools                      1.25-2.1                 amd64        tools for debugging the Intel graphics driver
ii  intel-media-va-driver-non-free:amd64 20.3.0+ds1-1             amd64        VAAPI driver for the Intel GEN8+ Graphics family
ii  intel-microcode                      3.20200616.1             amd64        Processor microcode firmware for Intel CPUs
ii  libdrm-intel1:amd64                  2.4.102-1                amd64        Userspace interface to intel-specific kernel DRM services -- runtime
ii  xserver-xorg-video-intel             2:2.99.917+git20200714-1 amd64        X.Org X server -- Intel i8xx, i9xx display driver
$ DISPLAY=:0 vainfo 
error: XDG_RUNTIME_DIR not set in the environment.
libva info: VA-API version 1.9.0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_9
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.9 (libva 2.9.0)
vainfo: Driver version: Intel iHD driver for Intel(R) Gen Graphics - 20.3.0 ()
vainfo: Supported profile and entrypoints
      VAProfileNone                   :	VAEntrypointVideoProc
      VAProfileNone                   :	VAEntrypointStats
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Simple            :	VAEntrypointEncSlice
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointFEI
      VAProfileH264Main               :	VAEntrypointEncSliceLP
      VAProfileH264High               :	VAEntrypointVLD
      VAProfileH264High               :	VAEntrypointEncSlice
      VAProfileH264High               :	VAEntrypointFEI
      VAProfileH264High               :	VAEntrypointEncSliceLP
      VAProfileVC1Simple              :	VAEntrypointVLD
      VAProfileVC1Main                :	VAEntrypointVLD
      VAProfileVC1Advanced            :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointEncPicture
      VAProfileH264ConstrainedBaseline:	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSlice
      VAProfileH264ConstrainedBaseline:	VAEntrypointFEI
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSliceLP
      VAProfileVP8Version0_3          :	VAEntrypointVLD
      VAProfileVP8Version0_3          :	VAEntrypointEncSlice
      VAProfileHEVCMain               :	VAEntrypointVLD
      VAProfileHEVCMain               :	VAEntrypointEncSlice
      VAProfileHEVCMain               :	VAEntrypointFEI
      VAProfileHEVCMain10             :	VAEntrypointVLD
      VAProfileHEVCMain10             :	VAEntrypointEncSlice
      VAProfileVP9Profile0            :	VAEntrypointVLD
      VAProfileVP9Profile2            :	VAEntrypointVLD
```

I have long had this in place:
```
cat /etc/X11/xorg.conf.d/20-intel.conf
Section "Device"
    Identifier  "Intel Graphics"
    Driver      "intel"
    Option      "TearFree" "true"
    Option      "AccelMethod"    "sna"
EndSection
```

I've forgotten to save a `~/.local/share/xorg/Xorg.0.log` from when things are going wrong, but from memory it contained all the expected modelines and indications it was getting EDD from the TV. I'll be sure to grab this next time.

**Comment by mark-kendall** [comment]
How are you starting mythfrontend? i.e. is this automatic at startup? if so, is it repeatable if you start mythfrontend manually?

Looking at the code and looking at your logs, I'm struggling to see what mythtv is or could be doing wrong. The xrandr calls are working and no errors are reported - they just seem to return a single modeline. My gut feeling is that this is not a mythtv issue - but obviously I can't be sure.

btw - if it is some sort of timing issue (e.g. frontend starts before display is turned on etc), then the fact that xrandr displays the expected output is probably a red herring - unless you can actually run xrandr at the same time.

**Comment by ijc** [comment]
Hi @mark-kendall, thanks for looking into this.

> How are you starting mythfrontend? i.e. is this automatic at startup? if so, is it repeatable if you start mythfrontend manually?

It's a bit oldschool (it's been this way since ~2006...): in `/etc/inittab` I have a line (re)spawning X via a script which calls `startx` as the mythtv user and then that user an `.xinitrc` which starts `evilwm` in the background and `mythfrontend` in the foreground (so exiting myth restarts the whole X server).

I've not tried starting myth manually but next time this starts happening I'll have xinitrc start an xterm instead of mythfrontend and try it. I suppose if that helps a short sleep in xinitrc might be worth a go too.

> Looking at the code and looking at your logs, I'm struggling to see what mythtv is or could be doing wrong. The xrandr calls are working and no errors are reported - they just seem to return a single modeline. My gut feeling is that this is not a mythtv issue - but obviously I can't be sure.

Yes, it does all look pretty sane on the mythtv end to me too. Am I reading correctly that mythfrontend queries the xrandr info once at startup and then when it is mode switching it is only looking through that set of modes rather than querying xrandr afresh?

This reminded me of another thing I tried at one point when this was happening which was pulling the HDMI cable and replugging it in the hopes it would redetect the right stuff, but unfortunately it didn't help.

One thing which lead me away from a timing issue was that things seem to be completely reliable after a reboot (gets all the right modes) but once it gets into this state later I've restarted the X stack a dozen or more times without ever succeeding -- if it was timing I'd expect it to work at least now and again I think? I suppose with hardware involved on the EDID side of things anything could be going on though to slow it down only from a certain point in time.

> btw - if it is some sort of timing issue (e.g. frontend starts before display is turned on etc), then the fact that xrandr displays the expected output is probably a red herring - unless you can actually run xrandr at the same time.

Agreed, although I've been restarting with the screen both off an on and it seems to make no difference.

In any case I've just dropped an `xrandr` call into `.xinitrc` right before mythfrontend is started.

BTW, I'm running Qt 5.15.1+dfsg-2 in case that is relevant.

**Comment by ijc** [comment]
Oooh, I wonder if the "wait for the backend..." holding screen thing I see on reboot is delaying the actual "real" start of the frontend enough to make it work? That doesn't happen when I just exit mythfrontend and restart X (or if it does happen it's too quick to see). That might explain why rebooting seems to fix it -- and in which case starting via xterm or after a sleep ought to be just as good.

**Comment by ijc** [comment]
I've not seen this problem since I reported it (typical! :roll_eyes:). On 2020-11-21 I did an `apt upgrade` maybe that fixed it (but I didn't see it in the intervening week between reporting and the upgrade), or maybe the `xrandr` I added to `.xinitrc` skews the timing just enough (or maybe I just had a run of bad/good luck).

At this point I guess there is minimal value in keep this open, if it does start happening again I do at least have some things to try before I reopen/report back.
